### PR TITLE
Override partial installs by default

### PR DESCRIPTION
**Pull request by scheibelp**
Fixes https://github.com/LLNL/spack/issues/2634

@tgamblin we originally talked about not automatically removing the prefix if --keep-prefix is set. While working on this I also noticed that it is possible that a partially complete installation may also have left stage files and in some cases re-initiating the install will fail because of that (gmp is an example as the 'configure' step cannot be performed twice without an intervening make clean). ~~My thought there was to also remove the stage files by default unless the user has specified --keep-stage. Perhaps adding a different option would be better.~~

(EDIT 3/27) --keep-prefix now indicates that the install prefix and staging directory should be kept if present. If --keep-prefix is not set, then both of these are removed.

- [x] Test removal of stage by default
- [x] Add test to ensure that the prefix is not deleted on a new install attempt if --keep-prefix is set

Depending on how spack is terminated in the middle of building a package it may leave a partially installed package in the install prefix. Originally Spack treated the package as being installed if the prefix was present, in which case the user would have to manually remove the installation prefix before restarting an install. This adds a more-thorough check to ensure that a package is actually installed. If the installation prefix is present but Spack determines that the install did not complete, it removes the installation prefix and starts a new install. ~~If the user has enabled --keep-prefix, then Spack reverts to its old behavior.~~

EDIT 3/27: If the user has enabled --keep-prefix, then Spack will keep its stage directory if present (as it did before); in this case Spack will also keep the installation prefix and re-execute the installation commands unless there is an issue with the metadata (e.g. the spec.yaml file), in which case it will reset the installation prefix.

### CODE REVIEW

**Comment by adamjstewart** [comment]
Do you think it would be worth wrapping #1915 into this or is it completely separate?

**Comment by scheibelp** [comment]
> Do you think it would be worth wrapping #1915 into this or is it completely separate?

I could take a crack at adding this in. Conceptually I think the intent is different. Also, I could perhaps take an approach that doesn't mess with the internals as much as was suggested in that PR; I think on how to improve this.

**Comment by scheibelp** [comment]
There is now support for the ```--force``` option with the ```install``` command: Spack will automatically remove the package and any of it's dependencies prior to installing it if it was already installed.

This is ready for review. If folks find it agreeable I'll want to rebase it to clean up the comment history but other than that I consider it complete

File: lib/spack/spack/cmd/install.py [hunk 1]
```diff
@@ -72,6 +72,9 @@ def setup_parser(subparser):
     subparser.add_argument(
         '--fake', action='store_true', dest='fake',
         help="fake install. just remove prefix and create a fake file")
+    subparser.add_argument(
+        '--force', '-f', action='store_true', dest='force',
```
Comment (adamjstewart) [inline lib/spack/spack/cmd/install.py]: If you're adding a new flag, can you update the tab-completion for `spack install`? See:
https://github.com/LLNL/spack/blob/develop/share/spack/spack-completion.bash#L408

**Comment by scheibelp** [comment]
> If you're adding a new flag, can you update the tab-completion for spack install

Done. Thanks for pointing that out!

Also I decided to remove the ```-f``` shorthand for ```--force```

**Comment by adamjstewart** [comment]
Why remove the shorthand? Most packages have one.

**Comment by scheibelp** [comment]
> Why remove the shorthand? Most commands have one.

5 of the 8 current options for the install command don't have a shorthand so I took that to mean I should decide whether it was a good idea. Since it is not entirely innocuous (e.g. if the user specifies it and cancels in the middle of an install they may be missing necessary packages for parents) I feel that keeping it ```--force``` forces the user to remember what it is and reduces the chance of accidentally using it.

**Comment by scheibelp** [comment]
Closing and reopening to run new tests that have been added (e.g. for python3)

**Pull request merged**